### PR TITLE
refactor: use let rec in `RedBlackTree`

### LINKED ISSUE #3288: let rec: refactor `RedBlackTree`

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #3288

### CODE REVIEW

**Comment by jaschdoc** [comment]
The compiler hangs on commit d09d15d due to the refactor on line 828.
```flix
case Node(color, left, key, v, right) => loop(left, kl -> loop(right, kr -> k(nodeA(color, kl, key, f(key, v), kr))))
```

**Comment by jaschdoc** [comment]
@magnus-madsen do you have any idea why it crashes as explained above?

**Comment by jaschdoc** [comment]
```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
	at ca.uwaterloo.flix.language.phase.unification.BoolUnification$.mkAnd(BoolUnification.scala:267)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:75)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:72)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:72)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:72)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:71)
	at ca.uwaterloo.flix.language.phase.unification.Substitution.visit$1(Substitution.scala:72)
```

**Comment by magnus-madsen** [comment]
Hopefully #3442 will fix this.

**Comment by jaschdoc** [comment]
It type checks now that #3555 has been merged.

File: main/src/library/RedBlackTree.flix [hunk 1]
```diff
@@ -121,207 +121,271 @@ namespace RedBlackTree {
...
     ///
     /// The function `f` must be pure.
     ///
-    @Time(time(f) * size(tree)) @Space(space(f) * Int32.log2(size(tree)))
-    pub def exists(f: (k, v) -> Bool, tree: RedBlackTree[k, v]): Bool = match tree {
-        case Node(_, a, k, v, b) => f(k, v) or exists(f, a) or exists(f, b)
-        case _ => false
-    }
+    @Time(time(f) * size(t)) @Space(space(f) * Int32.log2(size(t)))
+    pub def exists(f: (k, v) -> Bool, t: RedBlackTree[k, v]): Bool =
```
Comment (magnus-madsen) [inline main/src/library/RedBlackTree.flix:352]: For example, here the loop can also be omitted. Keep it simple stupid :)

File: main/src/library/RedBlackTree.flix [hunk 2]
```diff
@@ -121,207 +121,271 @@ namespace RedBlackTree {
...
-    @Time(Int32.log2(size(tree))) @Space(1)
-    pub def memberOf(k: k, tree: RedBlackTree[k, v]): Bool with Order[k] = match tree {
-        case Node(_, a, k1, _, b) =>
-            match k <=> k1 {
-                case LessThan => memberOf(k, a)
-                case EqualTo => true
-                case GreaterThan => memberOf(k, b)
+    @Time(Int32.log2(size(t))) @Space(1)
+    pub def memberOf(k: k, t: RedBlackTree[k, v]): Bool with Order[k] =
+        def loop(tt) = match tt {
```
[the remark this answers is not in the document]
Comment (jaschdoc) [reply main/src/library/RedBlackTree.flix:238]: Does the compiler inline `k` in this case? I'm just trying to reduce the number of variables that are copied at each recursive call

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/RedBlackTree.flix:238]: (However, using a loop will disable direct tail calls -- which is a different thing we would like to fix)

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/library/RedBlackTree.flix:238]: So I would just focus on the most natural style and then the compiler will do the right thing eventually.

**Comment by jaschdoc** [comment]
@magnus-madsen Ready for review

**Pull request merged**